design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,266 @@
+successfully by `ceph-volume`.
+
+The failure status of the last-applied `ceph-volume` command with the Drive Group spec is
+reported in `status`, and a corresponding `message` is also reported. The message must be filled in
+if the status is `false` (unsuccessful).
+
+Even though successful Drive Group applications are not reported `success` is still included in the
+status so that the failure is clear to human eyes.
+
+### Removing OSDs
+When removing an OSD (or OSDs) via the Rook Orchestrator Module, the Ceph mgr running in Kubernetes
+will also perform whatever Ceph operations are necessary to remove the OSD as would be expected by
+the user. This will replace the manual, imperative steps currently required by Rook-Ceph users to
+remove OSDs from the storage cluster documented
+[here for Rook v1.2](https://rook.io/docs/rook/v1.2/ceph-osd-mgmt.html#remove-an-osd).
+
+There is currently a proposal for a shared Ceph Mgr Orchestrator Module implementing common OSD
+removal steps [[link]](https://github.com/ceph/ceph/pull/32677) usable by any other Ceph Mgr
+Orchestrator module. The OSD removal mechanism should use this where possible.
+
+Rook will not automatically remove Kubernetes Deployments for OSDs unless
+`removeOSDsIfOutAndSafeToRemove: true`. This should be documented for users, and users should be
+advised to set this value to `true`.
+
+The `CephCluster` CRD allows users to add disks to nodes in several places including disks on all
+nodes in the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different
+ways). To avoid the complexity of having to manage the logic for determining how to remove a node
+from the `CephCluster` resource, OSD removal will **not** alter the `CephCluster` resource's
+`storage` specification with the exception of Drive Groups when needed. While modifying these items
+might be convenient for users, there are a lot of opportunities in development to miss corner cases
+or introduce logic bugs, which we should avoid.
+
+Users should be advised in documentation that if they plan to use the Ceph CLI or Dashboard to
+manage OSD addition in the Rook-Ceph cluster, it is recommended to set `useAllDevices: false`, leave
+both disk pattern matching empty, and to not specify any other disks in the `CephCluster` resource
+to avoid giving Rook the ability to re-add OSDs when the user does not wish it to. We should also
+note to users that Rook can automatically re-add removed OSDs if they are zapped manually by the
+user via documentation to avoid possible confusion and annoyance.
+
+#### Zapping OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section to allow OSDs to be zapped by Rook. Note that zapping is not the same as removing an OSD.
+Users are still expected to use either manual steps or the Ceph Orchestrator interfaces to remove
+OSDs from Ceph.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
```

Comment:
Rather than adding the zap feature to the CR, I was imagining that zapping would just be one more step to the imperative steps that will already need to be done with OSD removal. After removing the OSD, `c-v zap` could be called on the device with a job. Or why do you see it as a separate action?

Reply:
Maybe this would be best to discuss in the morning, but I am concerned that the mgr creating a job will require a "lock" mechanism on the OSDs as in my first design, and that seems ... prone to error. Unless that's not quite what you're proposing?

Reply:
I'm thinking we can avoid the locking with the following assumptions:
- Devices should be blacklisted/excluded from the driveGroups in the CR before the OSD is removed. OSDs would no longer be added to that device.
- The OSD daemon would be stopped before the OSD is removed
- If the admin is modifying the CR directly, they should ensure the device is not in their deviceFilter, driveGroups, or other settings (useAllDevices) that would cause the device to be added again.

Then there should not be contention for the disk. If there is still a random contention, the worst that happens is that the OSD will fail to start. We don't want it to start anyway.

Reply:
To your second point, this is only true if `removeOSDsIfOutAndSafeToRemove: true` is set. This could be something we have the mgr module be opinionated about.